### DISCUSSION TITLE
Not all tiles are rendering on Android with WebGL

### DISCUSSION THREAD

**Discussion by pdrhlik** [excaliburjs/Excalibur#2205] [Q&A]
I'm still very new to Excalibur but I managed to create a character and load a map though the Tiled plugin. The example map I'm using can be found [here](https://pipoya.itch.io/pipoya-rpg-tileset-32x32).

Everything is loaded perfectly in the browser and also on iOS. The issue is with Android. I get these 2 following errors there:
`WebGL: INVALID_VALUE: texImage2D: width or height out of range` and
`[.WebGL-0xdf10c600]RENDER WARNING: texture bound to texture unit 0 is not renderable. It might be non-power-of-2 or have incompatible texture filtering (maybe)?`

What happens is that it only loads some simple tiles but not tiles that are a bit more complicated and for example also contain some objects. I attached photos from all platforms to give you a better idea of what's happening.

Is this fixable or is it just an Android issue that we can't get past? I tried using the CanvasGraphicContext which works even on Android but it's basically unusable because of some lag.

Thanks in advance,
Patrik

**Android:**
<img width="300" alt="excalibur-web-screenshot" src="https://user-images.githubusercontent.com/14300398/151562313-45ccacc6-f4d1-4ab6-8d97-f3bb92bd935a.png" style="dsplay: inline;">

**iOS:**
<img width="300" alt="excalibur-web-screenshot" src="https://user-images.githubusercontent.com/14300398/151562333-acce1499-afc7-4b99-bbb8-e220a6cf764d.png" style="dsplay: inline;">

**Web:**
<img width="300" alt="excalibur-web-screenshot" src="https://user-images.githubusercontent.com/14300398/151562378-4bcdac1e-c271-4136-a71b-fcdb1f59f8a0.png" style="dsplay: inline;">

**Comment by eonarheim** [comment]
Hi @pdrhlik I'll take a look at this today, this definitely feels like an android related bug with the webgl code. 

I'm going to create a bug issue for this 👍

A few quick asks:
* What browser and version do you see this issue?
* What is the android OS version?
* What hardware do you see this issue on?
* Please send me some example images that aren't loading properly (it would be really awesome if you have a small reproduction repo for me)

I should be able to reproduce this in browserstack given enough info.

**Reply by pdrhlik** [reply]
Thanks for such a quick reply. A few quick answers for you:

- When running it on Android, I use Ionic Angular with Capacitor that handles the webview - It's probably using some Android Web view but I'm not sure what version that would be.
- Android 11
- Motorola G30 that's running Android One, meaning there it should be a pure android without any layers added by the manufacturer.
- I'll make a simple repo with this issue.

This is what it looks like on Android. Those black tiles should be rendered:
<img src="https://user-images.githubusercontent.com/14300398/151565995-c77ee7c8-2f0a-48cc-bf32-59074e615bae.png" width="300" />

**Reply by eonarheim** [reply]
Okay I think I know what is going on, on mobile devices they usually have a 4k limit per texture (4096x4096).

Excalibur will try to make textures power of two dimensions, so if the image atlas being uploaded exceeds 4096 in either width or height after that, it renders a black rectangle because the GPU refused the image. I've confirmed locally that if I have a (6912x96) image I see the same log output and behavior in browserstack that you see in your game.

Current ideas for fixes:
* Resize the texture on the fly to meet the GPU's reported maximum but with a console warning about the image size (this would be a lossy operation), but it would theoretically "just work" 😄 
* Throw a more friendly error explaining the image limitations and halt the game

Workaround:
* Resize any image assets below or to 4096x4096

**Reply by pdrhlik** [reply]
Well, I just finished the reproducible repo [https://github.com/pdrhlik/excalibur-android-tile-issue](https://github.com/pdrhlik/excalibur-android-tile-issue) if you want to have a look :-)

The map I'm using has several tilesets and one of them is 4256 high. I'll try and have a look at that. I haven't played with any maps or tilesets myself yet. I wanted to get comfortable with understanding how Excalibur works :-)

Anyway thanks a lot for your time!

**Reply by eonarheim** [reply]
Thanks for the detailed explanation!

I'll test any fixes against your repo to be extra sure we have a solution that catches everything (either with friendlier messaging or dynamic resizing).

**Reply by pdrhlik** [reply]
Perfect! Thanks as well. I can't wait to finish my first game :-)

**Reply by eonarheim** [reply]
@pdrhlik gh issue created https://github.com/excaliburjs/Excalibur/issues/2206 

I've got some code that adds some more helpful error messaging, not positive it's possible to resize on the fly easily.

Thoughts on the messaging?

Error situation (definitely won't render properly):
<img width="283" alt="image" src="https://user-images.githubusercontent.com/612071/151586686-bb6f4ac4-e2fc-426b-8455-40424d27b6c3.png">

I have a corresponding warning for images in general that are greater than 4096 in width/height so developers are warned in desktop environments:
<img width="231" alt="image" src="https://user-images.githubusercontent.com/612071/151586857-7538405a-417f-48e1-8f49-9e49c1adf3f4.png">
